onnxruntime/core/session/inference_session.cc

```diff
@@ -546,6 +546,48 @@ common::Status InferenceSession::InitializeSubgraphSessions(Graph& graph, Sessio
   return Status::OK();
 }
 
+static void ModelUseFP16Helper(const onnx::TypeProto type_proto, bool& use_fp16) {
+  switch (type_proto.value_case()) {
+    case ::onnx::TypeProto::ValueCase::kTensorType: {
+      if (type_proto.has_tensor_type()) {
+        auto& tensor_type = type_proto.tensor_type();
+        if (tensor_type.elem_type() == ONNX_NAMESPACE::TensorProto_DataType::TensorProto_DataType_FLOAT16) {
+          use_fp16 = true;
+          return;
+        }
+      }
+    } break;
+    case ::onnx::TypeProto::ValueCase::kSequenceType: {
+      if (type_proto.has_sequence_type()) {
+        auto& sequence_type = type_proto.sequence_type();
+        ModelUseFP16Helper(sequence_type.elem_type(), use_fp16);
+      }
+    } break;
+    case ::onnx::TypeProto::ValueCase::kMapType: {
+      if (type_proto.has_map_type()) {
+        auto& map_type = type_proto.map_type();
+        ModelUseFP16Helper(map_type.value_type(), use_fp16);
+      }
+    } break;
+  }
+}
+
+static bool ModelUseFP16(const onnx::ModelProto& model_proto) {
+  bool use_fp16 = false;
+  auto& graph = model_proto.graph();
+  auto& inputs = graph.input();
+  for (auto& input : inputs) {
+    if (input.has_name() && input.has_type()) {
+      auto& input_type_proto = input.type();
+      ModelUseFP16Helper(input_type_proto, use_fp16);
+      if (use_fp16) {
+        break;
+      }
+    }
+  }
+  return use_fp16;
+}
```

Comment:
WIth the changes to `ModelUseFP16Helper`, this could be something like:

```cpp
  const auto& graph = model_proto.graph();
  for (const auto& input : graph.input()) {
    if (input.has_name() && input.has_type() && ModelUseFP16Helper(input.type())) {
      return true;
    }
  }
  return false;
```

Or using functional C++, something like this (untested):

```cpp
  const auto& graph = model_proto.graph();
  const auto& inputs = graph.input();
  const auto uses_fp16 = [](const onnx::ValueInfoProto& input) {return input.has_name() && input.has_type() && ModelUseFP16Helper(input.type())};
  return std::any_of(std::begin(inputs), std::end(inputs), uses_fp16);
```